### PR TITLE
WIP: Implement  etcd, traefik and coredns templates

### PR DESCRIPTION
**Pull request by BolaNasr**

### CODE REVIEW

File: templates/coredns/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
+## template: github.com/threefoldtech/0-templates/coredns/0.0.1
+
+### Description:
+This is a node template responsible for managing [coredns](https://coredns.io/) server instance.
+
+### Schema:
+- `etcdServerName`: the name of etcd server that's created to get ip of it
```
Comment (zaibon) [inline templates/coredns/README.md]: `etcd` is enough
and the documentatio should say: `instance name of the etcd service used for configuration`

File: templates/coredns/coredns.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    @property
+    def _etcd(self):
+        return self.api.services.get(template_uid=ETCD_TEMPLATE_UID, name=self.data['etcdServerName'])
+    @property
+    def _etc_url(self):
+        result = self._etcd.schedule_action('connection_info').wait(die=True).result
+        return result['client_url']
+        
+    def node_port(self):
+        return self.data['nodePort']
```
Comment (zaibon) [inline templates/coredns/coredns.py]: this doesn't exist in the service schema.
also you should return the value from the coredns_sal

File: templates/coredns/coredns.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        self.state.set('status', 'running', 'ok')
+
+    def stop(self):
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('Stopping CoreDns %s' % self.name)
+        self._coredns_sal.stop()
+        self.state.delete('actions', 'start')
+        self.state.delete('status', 'running')
+
+    def register_domain(self, domain, ip, ttl=300):
```
Comment (zaibon) [inline templates/coredns/coredns.py]: I'm wondering if domain shoudn't be a service by iself instead of some methods here. @serboctor what do you think ?

File: templates/coredns/schema.capnp [hunk 4]
```diff
@@ -0,0 +1,34 @@
+@0xaf3b04a861b5a7f1;
+
+struct Schema {
+    etcdServerName @0: Text; #Name of Etcd service
```
Comment (zaibon) [inline templates/coredns/schema.capnp]: check my comment in the readme up

File: templates/coredns/schema.capnp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    upsteram @1 : Text; #by default 8.8.8.8:53 8.8.4.4:53
+    nodePort @2: Int32=9500; # the port to bind to
```
Comment (zaibon) [inline templates/coredns/schema.capnp]: should this be there|? cause it's not present in the documentation of the service

File: templates/etcd/etcd.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        self.state.delete('actions', 'install')
+        self.state.delete('status', 'running')
+
+    def connection_info(self):
+        self.state.check('status', 'running', 'ok')
+        etcd_sal = self._etcd_sal
+        return {'peer_url': etcd_sal.peer_url, 'client_url': etcd_sal.client_url}
+    
+    def update_cluster(self, cluster):
+        self.data['cluster'] = cluster
```
Comment (zaibon) [inline templates/etcd/etcd.py]: should restart the process too here

File: templates/etcd/etcd.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    
+    def insert_record(self, key, value):
```
Comment (zaibon) [inline templates/etcd/etcd.py]: these 3 methods doesn't belong here. I someone need to put record in etcd it needs to connect to it using a client

File: templates/minio/minio.py [hunk 8]
```diff
@@ -66,7 +66,6 @@ def node_port(self):
 
     def install(self):
         self.logger.info('Installing minio %s' % self.name)
-        minio_sal = self._minio_sal
```
Comment (zaibon) [inline templates/minio/minio.py]: why is this removed ?

File: templates/traefik/README.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
+## template: github.com/threefoldtech/0-templates/traefik/0.0.1
+
+### Description:
+This is a node template responsible for managing [traefik](https://docs.traefik.io/) server instance.
+
+### Schema:
+- `etcdServerName`: the name of etcd server that's created to get ip of it
```
Comment (zaibon) [inline templates/traefik/README.md]: same comment as the other

File: templates/traefik/schema.capnp [hunk 10]
```diff
@@ -0,0 +1,34 @@
+@0xb1e322af4c1787e4;
+
+struct Schema {
+    etcdServerName @0: Text; #Name of Etcd service
```
Comment (zaibon) [inline templates/traefik/schema.capnp]: same remark

File: templates/traefik/traefik.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        self.state.set('status', 'running', 'ok')
+
+    def stop(self):
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('Stopping traefik %s' % self.name)
+        self._traefik_sal.stop()
+        self.state.delete('actions', 'start')
+        self.state.delete('status', 'running')
+
+    def add_virtual_host(self, domain, ip, port=80):
```
Comment (zaibon) [inline templates/traefik/traefik.py]: 2 things:
1. I think virtual host should be a service by itself
2. you should not call actions on the etcd service to add records in etcd. instead get an etcd client and talk to etcd directly

File: templates/coredns/coredns.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+NODE_CLIENT = 'local'
+
+class Coredns(TemplateBase):
+    version = '0.0.1'
+    template_name = 'coredns'
+
+    def __init__(self, name, guid=None, data=None):
+        super().__init__(name=name, guid=guid, data=data)
+        self._url_ = None
```
[the remark this answers is not in the document]
Comment (BolaNasr) [reply templates/coredns/coredns.py]: i don't have it .. it's not my last commit

File: templates/coredns/coredns.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        self._node_ = None
+        self.etcd = self._etcd
```
[the remark this answers is not in the document]
Comment (BolaNasr) [reply templates/coredns/coredns.py]: it's not my last commit

File: templates/coredns/coredns.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            'recursive_resolvers':self.data['upsteram'],
+            'domain':self.data['domain'],
+            'etcd_endpoint': self.etc_url
+        }
+        return j.sal_zos.coredns.get(**kwargs)
+    @property
+    def _etcd(self):
+        return self.api.services.get(template_uid=ETCD_TEMPLATE_UID, name=self.data['etcdServerName'])
+
+    def etc_url(self):
```
[the remark this answers is not in the document]
Comment (BolaNasr) [reply templates/coredns/coredns.py]: i already i added it private method ...it's not my last commit

File: templates/coredns/coredns.py (hunk 3 above)
Comment (serboctor) [reply templates/coredns/coredns.py]: Hmm I dont think it is really worth it. But may be you're seeing something I am not. Why do you think it should be a service?

File: templates/coredns/schema.capnp (hunk 5 above)
Comment (serboctor) [reply templates/coredns/schema.capnp]: and in all cases the default is 53 not 9500 but yea this doesn't need to be here. Either you get the port from the sal, or in the case of dns it will always be 53.

File: templates/etcd/etcd.py (hunk 7 above)
Comment (serboctor) [reply templates/etcd/etcd.py]: I disagree. Wouldnt that mean duplicate code every where were we want to insert anything into etcd? what's the point of the services and sal if other services will connect using the client? Both coredns and traefik insert data into etcd and I don't see any reason why it shouldnt be done through the service.

File: templates/minio/minio.py (hunk 8 above)
Comment (serboctor) [reply templates/minio/minio.py]: Because it doesn't do anything. Doesn't even create the container.

File: templates/coredns/coredns.py (hunk 3 above)
Comment (zaibon) [reply templates/coredns/coredns.py]: Cause the coredns will most probably be a public service that can be used by different user.
But then, the domain service can be owned by different clients and they will own their domain when they created it.
If we don't have that separation, only the owner of the coredns service will be able to add domain to it

File: templates/etcd/etcd.py (hunk 7 above)
Comment (zaibon) [reply templates/etcd/etcd.py]: You are mixing stuff. A service is mean to del with the lifecycle of a application. It is not meant to provide an interface to talk to the application itself. 
And by having this here, we ARE duplicating code. While if you use the client, there is only one interface to talk to etcd, and it is using the client.

File: templates/coredns/coredns.py (hunk 3 above)
Comment (zaibon) [reply templates/coredns/coredns.py]: Ok so I've got more information about all this. It is another component that will connect to etcd and handle the management of the domain and virtual host. So there is no need to have such method on the services themself

File: templates/etcd/etcd.py (hunk 7 above)
Comment (zaibon) [reply templates/etcd/etcd.py]: Ok so I've got more information about all this. It is another component that will connect to etcd and handle the management of the domain and virtual host. So there is no need to have such method on the services themself
